File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/RequireHookStageRule.java
package ai.harness.maven;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.List;
import java.util.stream.Stream;

/**
 * Rule that requires hook files to contain stage markers.
 */
public class RequireHookStageRule implements HarnessCheckRule {
    private static final String CATEGORY = "requireHookStage";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public List<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        JsonNode catNode = manifest.get(CATEGORY);
        JsonNode stages = catNode.get("parameters").get("stages");
        JsonNode stackStages = stages.get("maven");
        if (stackStages == null) {
            return List.of();
        }
        String markerTemplate = catNode.get("parameters").get("markerTemplate").asText();
        String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return Stream.of(stackStages.fieldNames().spliterator(), false)
                .flatMap(hookName -> validateStage(root, hookName, stackStages.get(hookName).asText(), markerTemplate, severity).stream())
                .toList();
    }

    private List<Finding> validateStage(Path root, String hookName, String expectedStage, String markerTemplate, String severity) throws MojoExecutionException {
        String hookPath = "docs/harness/git-hooks/" + hookName;
        Path hook = root.resolve(hookPath);
        if (!HarnessCheckHelper.isSafeRegularFile(root, hook)) {
            return List.of();
        }
        String text = HarnessCheckHelper.readFile(root, hook);
        String expectedMarker = markerTemplate.replace("{stage}", expectedStage);
        return text.contains(expectedMarker)
                ? List.of()
                : List.of(new Finding(severity, CATEGORY, hookPath + " must contain stage marker '" + expectedMarker + "'"));
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/RequireHookStageRule.kt
package ai.harness.gradle

import kotlinx.serialization.json.JsonObject
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isRegularFile
import kotlin.io.path.readText

/**
 * Rule that requires hooks to contain stage markers.
 */
class RequireHookStageRule : HarnessCheckRule {
	override fun applies(manifest: JsonObject): Boolean {
		val category = "requireHookStage"
		val catObj = manifest[category]?.jsonObject ?: return false
		val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
		return enabled
	}

	override fun validate(manifest: JsonObject, root: Path, psiResults: HarnessPsiResults?): Collection<Finding> = buildSet {
		val category = "requireHookStage"
		val severity = HarnessCheck.Companion.severityOf(manifest, category)
		val catObj = manifest[category]?.jsonObject ?: return@buildSet
		val parametersObj = catObj["parameters"]?.jsonObject ?: return@buildSet
		val messagesObj = catObj["messages"]?.jsonObject ?: return@buildSet
		val markerTemplate = HarnessCheck.Companion.stringFrom(parametersObj, "markerTemplate")
		val stagesObj = parametersObj["stages"]?.jsonObject ?: return@buildSet
		val gradleStages = stagesObj["gradle"]?.jsonObject ?: return@buildSet
		val preCommitStage = HarnessCheck.Companion.stringFrom(gradleStages, "pre-commit")
		val prePushStage = HarnessCheck.Companion.stringFrom(gradleStages, "pre-push")

		return buildSet<Finding> {
			val preCommitHook = root / "docs/harness/git-hooks/pre-commit"
			if (preCommitHook.isRegularFile()) {
				val marker = markerTemplate.replace("{stage}", preCommitStage)
				if (!preCommitHook.readText().contains(marker)) {
					val msg = HarnessCheck.Companion.stringFrom(messagesObj, "default").takeIf { it.isNotEmpty() } ?: "pre-commit must contain stage marker '$marker'"
					add(Finding(severity, category, msg))
				}
			}

			val prePushHook = root / "docs/harness/git-hooks/pre-push"
			if (prePushHook.isRegularFile()) {
				val marker = markerTemplate.replace("{stage}", prePushStage)
				if (!prePushHook.readText().contains(marker)) {
					val msg = HarnessCheck.Companion.stringFrom(messagesObj, "default").takeIf { it.isNotEmpty() } ?: "pre-push must contain stage marker '$marker'"
					add(Finding(severity, category, msg))
				}
			}
		}.toList()
	}
}
